src/System.Text.Json/src/System/Text/Json/JsonReader.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Text.Parsing;
+using System.Text.Utf8;
+
+namespace System.Text.Json
+{
+    public struct JsonReader : IDisposable
+    {
+        private readonly Utf8String _str;
+        private int _index;
+
+        private static readonly List<Utf8CodeUnit> EmptyString = new List<Utf8CodeUnit>
+        {
+            new Utf8CodeUnit((byte) ' '),
+            new Utf8CodeUnit((byte) '\n'),
+            new Utf8CodeUnit((byte) '\r'),
+            new Utf8CodeUnit((byte) '\t')
+        };
+
+        private static readonly Utf8CodeUnit QuoteString = new Utf8CodeUnit((byte) '"');
+        private static readonly Utf8CodeUnit ColonString = new Utf8CodeUnit((byte) ':');
+        private static readonly Utf8CodeUnit CommaString = new Utf8CodeUnit((byte) ',');
+        private static readonly Utf8CodeUnit SquareOpenString = new Utf8CodeUnit((byte) '[');
+        private static readonly Utf8CodeUnit SquareCloseString = new Utf8CodeUnit((byte) ']');
+        private static readonly Utf8CodeUnit CurlyOpenString = new Utf8CodeUnit((byte) '{');
+        private static readonly Utf8CodeUnit CurlyCloseString = new Utf8CodeUnit((byte) '}');
+
+        public JsonReader(Utf8String str)
+        {
+            _str = str;
+            _index = 0;
+        }
+
+        public void Dispose()
+        {
+        }
+
+        public Utf8String ReadObjectStart()
+        {
+            SkipAll();
+            return ReadToByte(CurlyOpenString, true);
+        }
+
+        public Utf8String ReadObjectEnd()
+        {
+            SkipAll();
+            return ReadToByte(CurlyCloseString, true);
+        }
+
+        public Utf8String ReadArrayStart()
+        {
+            SkipAll();
+            return ReadToByte(SquareOpenString, true);
+        }
+
+        public Utf8String ReadArrayEnd()
+        {
+            SkipAll();
+            return ReadToByte(SquareCloseString, true);
+        }
+
+        public Utf8String ReadProperty()
+        {
+            SkipAll();
+            // TODO: Use _str.SubstringFrom(new Utf8String("\"")).Substring(1).SubstringTo(new Utf8String("\""));
+            ReadToByte(QuoteString, true);
+            return ReadToByte(QuoteString, false);
+        }
+
+        [CLSCompliant(false)]
+        public uint ReadPropertyValueInt()
```

Comment:
Does this cover all use case:
* ReadPropertyValueAsUInt (uses InvariantParser.TryParse(Utf8String, out uint, out int))
* ReadPropertyValueAsInt (reads possible "-" and calls ReadPropertyValueAsUInt)
* ReadPropertyValueAsLongUInt  (uses InvariantParser.TryParse(Utf8String, out ulong, out int))
* ReadPropertyValueAsLongInt (reads possible "-" and calls ReadPropertyValueAsLongUInt)

Along with ReadMemberValueAs* which just call ReadPropertyValueAs* (for arrays).